Mathlib/Tactic/Linter/HaveLetLinter.lean

```diff
@@ -92,26 +92,23 @@ def nonPropHaves : InfoTree → CommandElabM (Array (Syntax × Format)) :=
     let .original .. := stx.getHeadInfo | return #[]
     unless isHave? stx do return #[]
     let mctx := i.mctxAfter
-    let mvdecls := (i.goalsAfter.map (mctx.decls.find? ·)).reduceOption
-    -- we extract the `MetavarDecl` with largest index after a `have`, since this one
-    -- holds information about the metavariable where `have` introduces the new hypothesis.
-    let largestIdx := mvdecls.toArray.qsort (·.index > ·.index)
-    -- the relevant `LocalContext`
-    let lc := (largestIdx.getD 0 default).lctx
+    let mvdecls := i.goalsAfter.filterMap (mctx.decls.find? ·)
+    -- We extract the `MetavarDecl` with largest index after a `have`, since this one
+    -- holds information about the metavariable where `have` introduces the new hypothesis,
+    -- and determine the relevant `LocalContext`.
+    let lc := mvdecls.toArray.getMax? (·.index < ·.index) |>.getD default |>.lctx
     -- we also accumulate all `fvarId`s from all local contexts before the use of `have`
     -- so that we can then isolate the `fvarId`s that are created by `have`
-    let oldMvdecls := (i.goalsBefore.map (mctx.decls.find? ·)).reduceOption
-    let oldLctx := oldMvdecls.map (·.lctx)
-    let oldDecls := (oldLctx.map (·.decls.toList.reduceOption)).flatten
-    let oldFVars := oldDecls.map (·.fvarId)
-    -- `newDecls` are the local declarations whose `FVarID` did not exist before the `have`
-    -- effectively they are the declarations that we want to test for being in `Prop` or not.
+    let oldMvdecls := i.goalsBefore.filterMap (mctx.decls.find? ·)
+    let oldFVars := (oldMvdecls.map (·.lctx.decls.toList.reduceOption)).flatten.map (·.fvarId)
+    -- `newDecls` are the local declarations whose `FVarID` did not exist before the `have`.
+    -- Effectively they are the declarations that we want to test for being in `Prop` or not.
     let newDecls := lc.decls.toList.reduceOption.filter (! oldFVars.contains ·.fvarId)
-    -- now, we get the `MetaM` state up and running to find the types of each entry of `newDecls`
-    let fmts ← areProp_toFormat ctx lc (newDecls.map (·.type)).toArray
-    let (_propFmts, typeFmts) := (fmts.zip (newDecls.map (·.userName)).toArray).partition (·.1.1)
-    -- everything that is a Type triggers a warning on `have`
-    return typeFmts.map fun ((_, fmt), na) => (stx, f!"{na} : {fmt}"))
+    -- Now, we get the `MetaM` state up and running to find the types of each entry of `newDecls`.
+    -- For each entry which is a `Type`, we print a warning on `have`.
+    let fmts ← toFormat_propTypes ctx lc (newDecls.map (·.type)).toArray
+    let typeFmts := fmts.zip (newDecls.map (·.userName)).toArray
```

Comment:
This is wrong now, isn't it? You're zipping with something indexed by the original `newDecls`, but that needs to be filtered too. To get this right, we need to send the `userName` into `toFormat_propTypes` so it can be filtered there.